src/solver/conv_asm_implicit_gemm_v4r1_dynamic.cpp

```diff
@@ -0,0 +1,607 @@
+    }
+#endif
+    return 0;
+}
+
+bool PerformanceImplicitGemmV4R1Dynamic::
+operator==(const PerformanceImplicitGemmV4R1Dynamic& other) const
+{
+    // clang-format off
+    return BPerBlock == other.BPerBlock
+        && KPerBlock == other.KPerBlock
+        && EPerBlock == other.EPerBlock
+        && GemmNRepeat == other.GemmNRepeat
+        && GemmMPerThreadSubC == other.GemmMPerThreadSubC
+        && GemmNPerThreadSubC == other.GemmNPerThreadSubC
+        && GemmMLevel0Cluster == other.GemmMLevel0Cluster
+        && GemmNLevel0Cluster == other.GemmNLevel0Cluster
+        && GemmMLevel1Cluster == other.GemmMLevel1Cluster
+        && GemmNLevel1Cluster == other.GemmNLevel1Cluster
+        && InBlockCopyClusterLengths_E == other.InBlockCopyClusterLengths_E
+        && InBlockCopyClusterLengths_B == other.InBlockCopyClusterLengths_B
+        && InBlockCopyClusterLengths_N1 == other.InBlockCopyClusterLengths_N1
+        && InBlockCopyClusterLengths_N2 == other.InBlockCopyClusterLengths_N2
+        && WeiBlockCopyClusterLengths_E == other.WeiBlockCopyClusterLengths_E
+        && WeiBlockCopyClusterLengths_K == other.WeiBlockCopyClusterLengths_K;
+    // clang-format on
+}
+
+bool PerformanceImplicitGemmV4R1Dynamic::IsValid(const ConvolutionContext& ctx) const
+{
+    std::size_t N = KernelBatchN(ctx);
+    std::size_t K = KernelOutputChannelK(ctx);
+    std::size_t C = KernelInputChannelC(ctx);
+
+    std::size_t Ho = KernelOutputHeightHo(ctx);
+    std::size_t Wo = KernelOutputWidthWo(ctx);
+
+    std::size_t Y = KernelFilterHeightY(ctx);
+    std::size_t X = KernelFilterWidthX(ctx);
+
+    const int N1 = GemmNRepeat;
+    const int N2 = GemmNPerThreadSubC;
+    if(N % (N1 * N2) != 0)
+        return false; // wrong! cannot divice N evenly among thread
+
+    const auto N0 = N / (N1 * N2);
+
+    const auto B = N0 * Ho * Wo;
+
+    const auto nonVectorizedC = C / GetEPackLength(ctx, false);
+    const auto E              = nonVectorizedC * Y * X;
+
+    if(!(EPerBlock % InBlockCopyClusterLengths_E == 0 &&
+         EPerBlock % WeiBlockCopyClusterLengths_E == 0 &&
+         BPerBlock % InBlockCopyClusterLengths_B == 0 &&
+         KPerBlock % WeiBlockCopyClusterLengths_K == 0 && N1 % InBlockCopyClusterLengths_N1 == 0 &&
+         N2 % InBlockCopyClusterLengths_N2 == 0))
+        return false;
+
+    // divide block work by [K, B]
+    if(!(K % KPerBlock == 0 && B % BPerBlock == 0 && E % EPerBlock == 0))
+        return false; // wrong! cannot divice N evenly among thread
+
+    const auto KBlockWork = K / KPerBlock;
+    if(KBlockWork % ctx.group_counts != 0)
+        return false;
+
+    if((N1 * N2 * BPerBlock) % (GemmNPerThreadSubC * GemmNLevel0Cluster * GemmNLevel1Cluster) != 0)
+        return false;
+
+    // fp16/bfp16: doesn't support asymmetric matrix mul
+    if((ctx.IsFp16() || ctx.IsBfp16()) && GemmNPerThreadSubC != GemmMPerThreadSubC)
+        return false;
+
+    // fp16/bfp16: vector read of length 8 or greater is not supported
```

Comment:
Please explain (right in C++ comments) why you would like to keep this, or remove.